---
title: "Analysing Long Read RNA-Seq data with bambu"
author: Ying Chen, Yuk Kei Wan, Jonathan GÃ¶ke
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{bambu}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,tidy = TRUE,
  warning=FALSE, message=FALSE,
  comment = "##>"
)
```


# Introduction
*[Bambu](https://github.com/GoekeLab/bambu)* is a method  for transcript discovery and quantification from long read RNA-Seq data. Bambu uses aligned reads and genome reference annotations as input, and will return abundance estimates for all known transcripts and for newly discovered transcripts. Bambu uses the information from the reference annotations to correct misalignment at splice junctions, then reduces the aligned reads to read equivalent classes, and uses this information to identify novel transcripts across all samples of interest. Reads are then assigned to transcripts, and expression estimates are obtained using an expectation maximisation algorithm. Here, we present an example workflow for analysing Nanopore long read RNA-Sequencing data from two human cancer cell lines from the Singapore Nanopore Expression Project (SG-NEx).


# Content
* [Quick Start: Transcript discovery and quantification with bambu](#quick-start)
* [A complete workflow to identify and quantify transcript expression from Nanopore RNA-Seq data](#complete-workflow)
  + [Input data](#input-data)
    - [Aligned reads (bam files)](#bam-files)
    - [Genome sequence (fasta file/ BSGenome object)](#genome-sequences)
    - [Genome annotations (bambu annotations object/ gtf file / TxDb object)](#annotations)
  + [Transcript discovery and quantification](#transcript-discovery-quantification)
    - [Run bambu](#run-bambu)
    - [Visualise results](#visualise-results)
    - [Obtain gene expression estimates from transcript expression](#gene-expression)
    - [Save data (gtf/text)](#save-data)
  + [Identify differentially expressed genes](#DESeq2)
  + [Identify differential transcript usage](#DEXSeq)
* [Running bambu with large number of samples](#large-sample-num)
* [Getting Help](#get-help)
* [Citing bambu](#cite-bambu)
* [Session Information](#session-info)


# Quick start: Transcript discovery and quantification with bambu {#quick-start}

## Installation
You can install bambu from github:
```{r, eval = FALSE}
if (!requireNamespace("devtools", quietly = TRUE))
    install.packages("devtools")
devtools::install_github("GoekeLab/bambu")
```

## General Usage
The default mode to run *bambu* is using a set of aligned reads (bam files), reference genome annotations (gtf file, TxDb object, or bambuAnnotation object), and reference genome sequence (fasta file or BSgenome). bambu will return a summarizedExperiment object with the genomic coordinates for annotated and new transcripts and transcript expression estimates:
```{r, eval = FALSE}
library(bambu)
 
test.bam <- system.file("extdata", "SGNex_A549_directRNA_replicate5_run1_chr9_1_1000000.bam", package = "bambu")
 
se <- bambu(reads = test.bam, annotations = "TxDb.Hsapiens.UCSC.hg38.knownGene", genomeSequence = "BSgenome.Hsapiens.NCBI.GRCh38")
```

We highly recommend to use the same annotations that were used for genome alignment. If you have a gtf file and fasta file you can run bambu with the following options:
```{r, eval = FALSE}
test.bam <- system.file("extdata", "SGNex_A549_directRNA_replicate5_run1_chr9_1_1000000.bam", package = "bambu")
  
fa.file <- system.file("extdata", "Homo_sapiens.GRCh38.dna_sm.primary_assembly_chr9_1_1000000.fa", package = "bambu")
 
gtf.file <- system.file("extdata", "Homo_sapiens.GRCh38.91_chr9_1_1000000.gtf", package = "bambu")
 
bambuAnnotations <- prepareAnnotationsFromGTF(gtf.file)
 
se <- bambu(reads = test.bam, annotations = bambuAnnotations, genomeSequence = fa.file)
```
# A complete workflow to identify and quantify transcript expression from Nanopore RNA-Seq data {#complete-workflow}
To demonstrate the usage of Bambu, we used long-read RNA-Seq data generated using Oxford Nanopore Sequencing from the NanoporeRNASeq package, which consists of 6 samples from two human cell lines (K562 and MCF7). Each of these cell lines has three replicates, with 1 direct RNA sequencing data and 2 cDNA sequencing data. Reads are aligned to chromosome 22 (Grch38) and stored as bam files. We applied *bambu* to these bam files, visualized the results, and performed differential analysis to identify differentially expressed genes/transcripts. 

## Input data {#input-data}
### Aligned reads (bam files) {#bam-files}            
*bambu* takes genomic alignments saved in bam files. We used bamb-files from *NanoporeRNASeq* package, which contains reads aligned to the first half of the human chromosome 22 using *minimap2*. 
```{r}
devtools::load_all("/mnt/ont/s3.ontdata.store.genome.sg/projects/NanoporeRNASeq")
#devtools::install_github("GoekeLab/NanoporeRNASeq")
library(NanoporeRNASeq)
devtools::load_all()
library(bambu)

data("sample_info")
sample_info

data("bamFiles")
bamFiles <- Rsamtools::BamFileList(system.file("extdata",bamFiles,package = "NanoporeRNASeq"))
bamFiles
```

### Genome sequence (fasta file/ BSGenome object) {#genome-sequences}          
*bambu* additionally requires genome sequence, which is used to correct splicing junctions in read alignments. Ideally, we recommend the same genome seqeunce file that is used for alignment to be used for bambu. As an option, users can also choose to use BS genome object.
```{r}
# get path to fasta file
fasta_path <- system.file("extdata", "Homo_sapiens.GRCh38.dna_sm.primary_assembly_chr22_1_25409234.fa", package="NanoporeRNASeq")
# BSgenome
library(BSgenome.Hsapiens.NCBI.GRCh38)
```

### Genome annotations (bambu annotations object/ gtf file / TxDb object) {#annotations}    
*bambu* also requires annotation object, which is used in *bambu* to correct for start or end sites of matched read classes as well as determing the annotation type of identified transcripts. Here we used the annotation object from *NanoporeRNASeq*  package that is prepared by \code{prepareAnnotation} function in *bambu*.  
```{r}
data("annotation")
annotation
```

## Transcript discovery and quantification {#transcript-discovery-quantification}   
### Running bambu {#run-bambu}           
Applying *bambu* to perform isoform discovery to extend provided annotation, and then perform EM quantification on extended annotation.         
```{r, results = "hide"}
se <- bambu(reads = bamFiles, annotations = annotation, genomeSequence = "BSgenome.Hsapiens.NCBI.GRCh38", extendAnnotations = TRUE)
```
```{r}
se
```

Optionally, users can choose to apply *bambu* to do quantification without isoform discovery
```{r, results = "hide"}
seUnextended <- bambu(reads = bamFiles, annotations = annotation, genomeSequence = "BSgenome.Hsapiens.NCBI.GRCh38")
```
```{r}
seUnextended
```


### Visualise results {#visualise-results}               
*bambu* provides quick visualisation of transcript estimation results with \code{plot.bambu} function. With multiple samples, we can visualise the correlation clustering between cell lines with heatmap:
```{r, fig.width = 8, fig.height = 6}
colData(se)$groupVar <- sample_info$cellLine
colnames(se) <- sample_info$sample_id
colData(se)$name <- sample_info$sample_id
plot.bambu(se, group.variable = "groupVar", type = "heatmap")
```

Additionally, we can also visualise the correlation with a 2-dimmension PCA plot.
```{r, fig.width = 8, fig.height = 6}
colData(se)$groupVar <-  sample_info$cellLine
plot.bambu(se, group.variable = "groupVar", type = "pca")
```

In addition to visualising correlation between samples, we also provide quick look into the extended annotation by looking at gene examples. Here we looked at gene ENSG00000099968 and visualised the genomic annotations for annotated and novel isoforms for this gene and compared the expression levels for these isoforms across samples. 
```{r, fig.width = 8, fig.height = 10}
plot.bambu(se, type = "annotation", gene_id = "ENSG00000099968")

```

### Obtain gene expression estimates from transcript expression {#gene-expression}
Gene expression can be obtained from transcript expression obtained from *bambu* using this function. Looking at the output, we can see there are novel genes identified as well
```{r}
seGene <- transcriptToGeneExpression(se)
seGene
```

We compared the gene expression across the 6 samples. Generally, samples from the cell line showed higher correlation than across the cell lines for gene expression.
```{r, fig.width = 8, fig.height = 6}
colData(seGene)$groupVar <-  sample_info$cellLine
plot.bambu(seGene, group.variable = "groupVar", type = "heatmap")
```

Similarly, when we visualise using PCA plot, we see that samples for the same cell line are more clustered.
```{r, fig.width = 8, fig.height = 6}
colData(seGene)$groupVar <-  sample_info$cellLine
plot.bambu(seGene, group.variable = "groupVar", type = "pca")
```

### Save data (gtf/text){#save-data}
*bambu* allows write annotation output to gtf file and gene and transcript estimates to txt files.
```{r}
writeBambuOutput(se, path = "./bambu/")
```

## Identifying differentially expressed genes {#DESeq2}
We used *DESeq2* to find the differentially expressed genes. Similar to using results from *Salmon*, estimates from *bambu* will first be rounded.  
```{r}
library(DESeq2)
dds <- DESeqDataSetFromMatrix(apply(assays(seGene)$counts,c(1,2),round),
                                   colData = colData(se),
                                   design = ~ groupVar)
dds.deseq <- DESeq(dds)
deGeneRes <- DESeq2::results(dds.deseq, independentFiltering=FALSE)
head(deGeneRes[order(deGeneRes$padj),])
```
A quick summary of differentially expressed genes
```{r}
summary(deGeneRes)
```

We can also visualise the MA-plot for differentially used isoforms
```{r, fig.width = 8, fig.height = 6}
plotMA(deGeneRes, ylim = c(-3,3))
```

However, as recommended by [*DESeq2* tutorial](http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#alternative-shrinkage-estimators), visualizing the MA-plots using the original log-fold change results will be affected by the noise associated with log2 fold changes from low count genes without requiring arbitrary filtering thresholds. So we applied same shrinkage to effect sizes to improve the visualization.
```{r, fig.width = 8, fig.height = 6}
library(apeglm)
resLFC <- lfcShrink(dds.deseq, coef="groupVar_MCF7_vs_K562", type="apeglm")
plotMA(resLFC, ylim = c(-3,3))
```

## Identifying differential expression for isoform detection {#DEXSeq}
We used *DEXSeq* to detect alternative used isoforms.
```{r}
count.data <- cbind(data.frame(gene_id = as.data.frame(rowData(se))$GENEID,
                               feature_id = as.data.frame(rowData(se))$TXNAME),
                               assays(se)$counts)
```

```{r}
library(DEXSeq)
formula <- as.formula("~sample + exon + groupVar:exon")
dxd <- DEXSeqDataSet(countData=round(as.matrix(count.data[,-c(1:2)])),
                     sampleData=as.data.frame(colData(se)),
                     design=formula,
                     featureID = count.data$feature_id,
                     groupID=count.data$gene_id)
dxd <- estimateExonFoldChanges(testForDEU(estimateDispersions(estimateSizeFactors(dxd), formula = formula), fullModel = formula), fitExpToVar="groupVar")

dxr <- DEXSeqResults(dxd, independentFiltering=FALSE)
head(dxr)
```
We can visualize the MA-plot
```{r,fig.width = 8, fig.height = 6}
plotMA(dxr, cex=0.8 )
```



# Running bambu with large number of samples {#large-sample-num}
For larger sample numbers we recommend to write the processed data to a file:
```{r, eval = FALSE}
bambu(reads = test.bam, readClass.outputDir = "./bambu/", annotations = bambuAnnotations, genomeSequence = fa.file)
```
# Getting Help {#get-help}
## Advanced Options
More stringent filtering thresholds imposed on potential novel transcripts    

* Keep novel transcripts with min 5 read count in at least 1 sample:
```{r, eval = FALSE}
bambu(reads, annotations, genomeSequence, isoreParameters = list(min.readCount = 5))
```
* Keep novel transcripts with min 5 samples having at least 2 counts:
```{r, eval = FALSE}
bambu(reads, annotations, genomeSequence, isoreParameters = list(min.sampleNumber = 5))
```
* Filter out transcripts with relative abundance within gene lower than 10%:
```{r, eval = FALSE}
bambu(reads, annotations, genomeSequence, isoreParameters = list(min.readFractionByGene = 0.1))
```
Quantification without bias correction          
The default estimation automatically does bias correction for expression estimates. However, you can choose to perform the quantification without bias correction.
```{r, eval = FALSE}
bambu(reads, annotations, genomeSequence, emParameters = list(bias = FALSE))
```
Parallel computation       
bambu allows parallel computation.
```{r, eval = FALSE}
bambu(reads, annotations, genomeSequence, ncore = 8)
```
See *[manual](https://github.com/GoekeLab/bambu/blob/master/docs/bambu_0.1.0.pdf)* for details to customize other conditions.


# Citing bambu {#cite-bambu}
A manuscript describing bambu is currently in preparation. If you use bambu for your research, please cite using the following doi: 10.5281/zenodo.3900025.

# Session Information {#session-info}
```{r}
sessionInfo()
```



