<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Reference-guided isoform reconstruction and quantification for long read RNA-Seq data • bambu</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Reference-guided isoform reconstruction and quantification for long read RNA-Seq data">
<meta property="og:description" content="Multi-sample transcript discovery and quantification using long read RNA-Seq data. ">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">bambu</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="articles/bambu.html">Get started</a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/GoekeLab/bambu/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">


<div id="bambu-reference-guided-transcript-discovery-and-quantification-for-long-read-rna-seq-data" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#bambu-reference-guided-transcript-discovery-and-quantification-for-long-read-rna-seq-data" class="anchor"></a>bambu: reference-guided transcript discovery and quantification for long read RNA-Seq data</h1></div>
<p><strong><em>bambu</em></strong> is a R package for multi-sample transcript discovery and quantification using long read RNA-Seq data. You can use <strong><em>bambu</em></strong> after read alignment to obtain expression estimates for known and novel transcripts and genes. The output from <strong><em>bambu</em></strong> can directly be used for visualisation and downstream analysis such as differential gene expression or transcript usage.</p>
<p><a href="https://github.com/GoekeLab/bambu"><img src="https://img.shields.io/github/v/release/GoekeLab/bambu?style=plastic" alt="GitHub release (latest by date)"></a> <a href="https://github.com/GoekeLab/bambu/graphs/contributors"><img src="https://img.shields.io/badge/Maintained%3F-Yes-blue" alt="Maintained?"></a><br><a href="#installation"><img src="https://img.shields.io/badge/Install-Github-brightgreen" alt="Install"></a> <a href="https://www.gnu.org/licenses/gpl-3.0"><img src="https://img.shields.io/badge/License-GPLv3-blue.svg" alt="License: GPL v3"></a></p>
<hr>
<div id="content" class="section level3">
<h3 class="hasAnchor">
<a href="#content" class="anchor"></a>Content</h3>
<ul>
<li><a href="#installation">Installation</a></li>
<li><a href="#general-usage">General usage</a></li>
<li><a href="#use-precalculated-annotation-objects">Use precalculated annotations object</a></li>
<li><a href="#advanced-options">Advanced options</a></li>
<li><a href="#complementary-functions">Complementary functions</a></li>
<li><a href="#contributors">Contributors</a></li>
</ul>
</div>
<div id="installation" class="section level3">
<h3 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h3>
<p>You can install <strong><em>bambu</em></strong> from github:</p>
<pre class="rscript"><code>if (!requireNamespace("devtools", quietly = TRUE))
    install.packages("devtools")
devtools::install_github("GoekeLab/bambu")</code></pre>
<hr>
</div>
<div id="general-usage" class="section level3">
<h3 class="hasAnchor">
<a href="#general-usage" class="anchor"></a>General Usage</h3>
<p>The default mode to run <em><strong>bambu</strong> is using a set of aligned reads (bam files), reference genome annotations (gtf file, TxDb object, or bambuAnnotation object), and reference genome sequence (fasta file or BSgenome). </em><strong>bambu</strong>* will return a summarizedExperiment object with the genomic coordinates for annotated and new transcripts and transcript expression estimates:</p>
<p>```rscript library(bambu)</p>
<p>test.bam &lt;- system.file(“extdata”, “SGNex_A549_directRNA_replicate5_run1_chr9_1_1000000.bam”, package = “bambu”)</p>
<p>se &lt;- bambu(reads = test.bam, annotations = “TxDb.Hsapiens.UCSC.hg38.knownGene”, genomeSequence = “BSgenome.Hsapiens.NCBI.GRCh38”)</p>
<p>```</p>
<p>We highly recommend to use the same annotations that were used for genome alignment. If you have a gtf file and fasta file you can run <strong><em>bambu</em></strong> with the following options:</p>
<pre class="rscript"><code>test.bam &lt;- system.file("extdata", "SGNex_A549_directRNA_replicate5_run1_chr9_1_1000000.bam", package = "bambu")
  
fa.file &lt;- system.file("extdata", "Homo_sapiens.GRCh38.dna_sm.primary_assembly_chr9_1_1000000.fa", package = "bambu")

gtf.file &lt;- system.file("extdata", "Homo_sapiens.GRCh38.91_chr9_1_1000000.gtf", package = "bambu")

bambuAnnotations &lt;- prepareAnnotationsFromGTF(gtf.file)

se &lt;- bambu(reads = test.bam, annotations = bambuAnnotations, genomeSequence = fa.file)
</code></pre>
<p><strong>Quantification of annotated transcripts and genes only (no transcript/gene discovery)</strong></p>
<pre class="rscript"><code><a href="reference/bambu.html">bambu(reads = test.bam, annotations = txdb, genomeSequence = fa.file, extendAnnotations = FALSE)</a></code></pre>
<p><strong>Large sample number/ limited memory</strong><br>
For larger sample numbers we recommend to write the processed data to a file:</p>
<pre class="rscript"><code><a href="reference/bambu.html">bambu(reads = test.bam, readClass.outputDir = "./bambu/", annotations = bambuAnnotations, genomeSequence = fa.file)</a></code></pre>
<hr>
</div>
<div id="use-precalculated-annotation-objects" class="section level3">
<h3 class="hasAnchor">
<a href="#use-precalculated-annotation-objects" class="anchor"></a>Use precalculated annotation objects</h3>
<p>You can also use precalculated annotations.</p>
<p>If you plan to run <strong><em>bambu</em></strong> more frequently, we recommend to save the bambuAnnotations object.</p>
<p>The bambuAnnotation object can be calculated from a gtf file:</p>
<pre class="rscript"><code>annotations &lt;- prepareAnnotationFromGTF(gtf.file)</code></pre>
<p>From TxDb object</p>
<pre class="rscript"><code>annotations &lt;- prepareAnnotations(txdb)</code></pre>
<hr>
</div>
<div id="advanced-options" class="section level2">
<h2 class="hasAnchor">
<a href="#advanced-options" class="anchor"></a>Advanced Options</h2>
<p><strong>More stringent filtering thresholds imposed on potential novel transcripts</strong></p>
<blockquote>
<p>Keep novel transcripts with min 5 read count in at least 1 sample:</p>
</blockquote>
<pre class="rscript"><code><a href="reference/bambu.html">bambu(reads, annotations, genomeSequence, isoreParameters = list(min.readCount = 5))</a></code></pre>
<blockquote>
<p>Keep novel transcripts with min 5 samples having at least 2 counts:</p>
</blockquote>
<pre class="rscript"><code><a href="reference/bambu.html">bambu(reads, annotations, genomeSequence, isoreParameters = list(min.sampleNumber = 5))</a></code></pre>
<blockquote>
<p>Filter out transcripts with relative abundance within gene lower than 10%:</p>
</blockquote>
<pre class="rscript"><code><a href="reference/bambu.html">bambu(reads, annotations, genomeSequence, isoreParameters = list(min.readFractionByGene = 0.1))</a></code></pre>
<p><strong>Quantification without bias correction</strong></p>
<blockquote>
<p>The default estimation automatically does bias correction for expression estimates. However, you can choose to perform the quantification without bias correction.</p>
</blockquote>
<pre class="rscript"><code><a href="reference/bambu.html">bambu(reads, annotations, genomeSequence, emParameters(bias = FALSE))</a></code></pre>
<p><strong>Parallel computation</strong></p>
<blockquote>
<p><strong><em>bambu</em></strong> allows parallel computation.</p>
</blockquote>
<pre class="rscript"><code><a href="reference/bambu.html">bambu(reads, annotations, genomeSequence, ncore = 8)</a></code></pre>
<p>See <a href="docs/bambu_0.1.0.pdf">manual</a> for details to customize other conditions.</p>
<hr>
<div id="complementary-functions" class="section level3">
<h3 class="hasAnchor">
<a href="#complementary-functions" class="anchor"></a>Complementary functions</h3>
<p><strong>Transcript expression to gene expression</strong></p>
<pre class="rscript"><code><a href="reference/transcriptToGeneExpression.html">transcriptToGeneExpression(se)</a></code></pre>
<p><strong>Visualization</strong></p>
<blockquote>
<p>You can visualize the novel genes/transcripts using <strong><em>plot.bambu</em></strong> function</p>
</blockquote>
<pre class="rscript"><code><a href="reference/plot.bambu.html">plot.bambu(se, type = "annotation", gene_id)

plot.bambu(se, type = "annotation", transcript_id)</a></code></pre>
<blockquote>
<p><strong><em>plot.bambu</em></strong> can also be used to visualize the clustering of input samples on gene/transcript expressions</p>
</blockquote>
<pre class="rscript"><code><a href="reference/plot.bambu.html">plot.bambu(se, type = "heatmap") # heatmap 

plot.bambu(se, type = "pca") # PCA visualization</a></code></pre>
<blockquote>
<p><strong><em>plot.bambu</em></strong> can also be used to visualize the clustering of input samples on gene/transcript expressions with grouping variable</p>
</blockquote>
<pre class="rscript"><code><a href="reference/plot.bambu.html">plot.bambu(se, type = "heatmap", group.var) # heatmap 

plot.bambu(se, type = "pca", group.var) # PCA visualization</a></code></pre>
<p><strong>Write bambu outputs to files</strong></p>
<blockquote>
<p><strong><em>writeBambuOutput</em></strong> will generate three files, including a <strong>.gtf</strong> file for the extended annotations, and two <strong>.txt</strong> files for the expression counts at transcript and gene levels.</p>
</blockquote>
<pre class="rscript"><code><a href="reference/writeBambuOutput.html">writeBambuOutput(se, path = "./bambu/")</a></code></pre>
<hr>
</div>
<div id="contributors" class="section level3">
<h3 class="hasAnchor">
<a href="#contributors" class="anchor"></a>Contributors</h3>
<p>This package is developed and maintained by<a href="https://github.com/cying111">Ying Chen</a>, <a href="https://github.com/yuukiiwa">Yuk Kei Wan</a>, <a href="https://github.com/jonathangoeke">Jonathan Goeke</a> and at Genome Institute of Singapore. If you want to contribute, please leave an issue. Thank you.</p>
</div>
</div>
</div>

  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2>Links</h2>
<ul class="list-unstyled">
<li>Browse source code at <br><a href="https://github.com/GoekeLab/bambu/">https://​github.com/​GoekeLab/​bambu/​</a>
</li>
</ul>
</div>
<div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li>
<a href="https://www.r-project.org/Licenses/GPL-3">GPL-3</a> </li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>Ying Chen Developer <br><small class="roles"> Maintainer </small>  </li>
<li>Jonathan Goeke Developer <br><small class="roles"> Author </small>  </li>
</ul>
</div>

  <div class="dev-status">
<h2>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://raw.githubusercontent.com/GoekeLab/bambu/master/figures/transparent-bambu.png?token=AGA7DTCQ2VT5ILG3R6ORKUK6WP424"><img src="https://raw.githubusercontent.com/GoekeLab/bambu/master/figures/transparent-bambu.png?token=AGA7DTCQ2VT5ILG3R6ORKUK6WP424" title="Bambu" alt="Bambu"></a></li>
</ul>
</div>
</div>
</div>


      <footer><div class="copyright">
  <p>Developed by Ying Chen Developer, Jonathan Goeke Developer.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
