<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>bambu â€¢ bambu</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="bambu">
<meta property="og:description" content="bambu">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">bambu</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/bambu.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/GoekeLab/bambu/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>bambu</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/GoekeLab/bambu/blob/master/vignettes/bambu.Rmd"><code>vignettes/bambu.Rmd</code></a></small>
      <div class="hidden name"><code>bambu.Rmd</code></div>

    </div>

    
    
<div id="introduction" class="section level2">
<h2 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h2>
<p><em><a href="https://github.com/GoekeLab/bambu">Bambu</a></em> can be used for transcript discovery and quantification from long read RNA-Seq data. Here, we present an example usage of bambu on Nanopore long read RNA-Sequencing from 2 human cancer cell lines</p>
</div>
<div id="data-description" class="section level2">
<h2 class="hasAnchor">
<a href="#data-description" class="anchor"></a>Data description</h2>
<p>To demonstrate the usage of <em>Bambu</em>, we used the data in NanoporeRNASeq, which contains single chromosome RNA-Seq data from two common cell lines K562 and MCF7. Each of these cell line has three replicates, with 1 direct RNA sequencing data and 2 cDNA sequencing data.</p>
<div class="sourceCode"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">NanoporeRNASeq</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">bambu</span>)</pre></body></html></div>
<div class="sourceCode"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span>(<span class="st">"sample_info"</span>)
<span class="no">sample_info</span>
<span class="co">##&gt;                                sample_id Platform cellLine   protocol</span>
<span class="co">##&gt; 1: SGNex_K562_directcDNA_replicate1_run2   MinION     K562 directcDNA</span>
<span class="co">##&gt; 2: SGNex_K562_directcDNA_replicate4_run2  GridION     K562 directcDNA</span>
<span class="co">##&gt; 3:  SGNex_K562_directRNA_replicate6_run1  GridION     K562  directRNA</span>
<span class="co">##&gt; 4: SGNex_MCF7_directcDNA_replicate1_run2   MinION     MCF7 directcDNA</span>
<span class="co">##&gt; 5: SGNex_MCF7_directcDNA_replicate3_run3  GridION     MCF7 directcDNA</span>
<span class="co">##&gt; 6:  SGNex_MCF7_directRNA_replicate4_run1  GridION     MCF7  directRNA</span>
<span class="co">##&gt;        bioRep cancer_type</span>
<span class="co">##&gt; 1: replicate1   Leukocyte</span>
<span class="co">##&gt; 2: replicate4   Leukocyte</span>
<span class="co">##&gt; 3: replicate6   Leukocyte</span>
<span class="co">##&gt; 4: replicate1      Breast</span>
<span class="co">##&gt; 5: replicate3      Breast</span>
<span class="co">##&gt; 6: replicate4      Breast</span></pre></body></html></div>
<p>Name of bamfiles can be loaded as follows</p>
<div class="sourceCode"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span>(<span class="st">"bamFileNames"</span>)
<span class="no">bamFileNames</span>
<span class="co">##&gt; [1] "SGNex_K562_directcDNA_replicate1_run2_genome_chr22_1_25409234.bam"</span>
<span class="co">##&gt; [2] "SGNex_K562_directcDNA_replicate4_run2_genome_chr22_1_25409234.bam"</span>
<span class="co">##&gt; [3] "SGNex_K562_directRNA_replicate6_run1_genome_chr22_1_25409234.bam" </span>
<span class="co">##&gt; [4] "SGNex_MCF7_directcDNA_replicate1_run2_genome_chr22_1_25409234.bam"</span>
<span class="co">##&gt; [5] "SGNex_MCF7_directcDNA_replicate3_run3_genome_chr22_1_25409234.bam"</span>
<span class="co">##&gt; [6] "SGNex_MCF7_directRNA_replicate4_run1_genome_chr22_1_25409234.bam"</span></pre></body></html></div>
<p>We then loaded bam.files.</p>
<div class="sourceCode"><html><body><pre class="r"><span class="no">bam.file</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"extdata"</span>,<span class="no">bamFileNames</span>, <span class="kw">package</span> <span class="kw">=</span> <span class="st">"NanoporeRNASeq"</span>)
<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span>(<span class="st">"annotationGrangesList_chr22_1_25409234"</span>)</pre></body></html></div>
<p>We applied <em>bambu</em> to perform EM on extended annotations</p>
<div class="sourceCode"><html><body><pre class="r">seExtended &lt;- bambu(reads = bam.file, annotations = annotationGrangesList_chr22_1_25409234, genomeSequence = "BSgenome.Hsapiens.NCBI.GRCh38", extendAnnotations = TRUE, verbose = FALSE, ncore = 6)
##&gt; 
  |                                                                            
  |                                                                      |   0%
  |                                                                            
  |============                                                          |  17%
  |                                                                            
  |=======================                                               |  33%
  |                                                                            
  |===================================                                   |  50%
  |                                                                            
  |===============================================                       |  67%
  |                                                                            
  |==========================================================            |  83%
  |                                                                            
  |======================================================================| 100%
##&gt; 
##&gt; 
  |                                                                            
  |                                                                      |   0%
  |                                                                            
  |============                                                          |  17%
  |                                                                            
  |=======================                                               |  33%
  |                                                                            
  |===================================                                   |  50%
  |                                                                            
  |===============================================                       |  67%
  |                                                                            
  |==========================================================            |  83%
  |                                                                            
  |======================================================================| 100%
seExtended
##&gt; class: RangedSummarizedExperiment 
##&gt; dim: 1923 6 
##&gt; metadata(0):
##&gt; assays(2): counts CPM
##&gt; rownames(1923): tx.1 tx.2 ... ENST00000484509 ENST00000468442
##&gt; rowData names(4): TXNAME GENEID eqClass newTxClass
##&gt; colnames(6):
##&gt;   SGNex_K562_directcDNA_replicate1_run2_genome_chr22_1_25409234
##&gt;   SGNex_K562_directcDNA_replicate4_run2_genome_chr22_1_25409234 ...
##&gt;   SGNex_MCF7_directcDNA_replicate3_run3_genome_chr22_1_25409234
##&gt;   SGNex_MCF7_directRNA_replicate4_run1_genome_chr22_1_25409234
##&gt; colData names(1): name</pre></body></html></div>
<p><em>bambu</em> allows quantification without isoform discovery</p>
<div class="sourceCode"><html><body><pre class="r">se &lt;- bambu(reads = bam.file, annotations = annotationGrangesList_chr22_1_25409234, genomeSequence = "BSgenome.Hsapiens.NCBI.GRCh38", extendAnnotations = FALSE, verbose = FALSE, ncore = 6)
##&gt; 
  |                                                                            
  |                                                                      |   0%
  |                                                                            
  |============                                                          |  17%
  |                                                                            
  |=======================                                               |  33%
  |                                                                            
  |===================================                                   |  50%
  |                                                                            
  |===============================================                       |  67%
  |                                                                            
  |==========================================================            |  83%
  |                                                                            
  |======================================================================| 100%
##&gt; 
##&gt; 
  |                                                                            
  |                                                                      |   0%
  |                                                                            
  |============                                                          |  17%
  |                                                                            
  |=======================                                               |  33%
  |                                                                            
  |===================================                                   |  50%
  |                                                                            
  |===============================================                       |  67%
  |                                                                            
  |==========================================================            |  83%
  |                                                                            
  |======================================================================| 100%
se
##&gt; class: RangedSummarizedExperiment 
##&gt; dim: 1500 6 
##&gt; metadata(0):
##&gt; assays(2): counts CPM
##&gt; rownames(1500): ENST00000624155 ENST00000422332 ... ENST00000484509
##&gt;   ENST00000468442
##&gt; rowData names(3): TXNAME GENEID eqClass
##&gt; colnames(6):
##&gt;   SGNex_K562_directcDNA_replicate1_run2_genome_chr22_1_25409234
##&gt;   SGNex_K562_directcDNA_replicate4_run2_genome_chr22_1_25409234 ...
##&gt;   SGNex_MCF7_directcDNA_replicate3_run3_genome_chr22_1_25409234
##&gt;   SGNex_MCF7_directRNA_replicate4_run1_genome_chr22_1_25409234
##&gt; colData names(1): name</pre></body></html></div>
</div>
<div id="check-expression-estimates" class="section level2">
<h2 class="hasAnchor">
<a href="#check-expression-estimates" class="anchor"></a>Check expression estimates</h2>
<p>We can check the estimated transcript expression using heatmap:</p>
<div class="sourceCode"><html><body><pre class="r"><span class="fu">colData</span>(<span class="no">seExtended</span>)$<span class="no">groupVar</span> <span class="kw">&lt;-</span>  <span class="fu"><a href="https://rdrr.io/r/base/unlist.html">unlist</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(<span class="no">seExtended</span>),<span class="kw">function</span>(<span class="no">x</span>) <span class="fu"><a href="https://rdrr.io/r/base/unlist.html">unlist</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/strsplit.html">strsplit</a></span>(<span class="no">x</span>,<span class="st">"_"</span>))[<span class="fl">2</span>]))
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(<span class="no">seExtended</span>) <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">gsub</a></span>(<span class="st">"_genome_chr22_1_25409234"</span>,<span class="st">""</span>,<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(<span class="no">seExtended</span>))
<span class="fu">colData</span>(<span class="no">seExtended</span>)$<span class="no">name</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">gsub</a></span>(<span class="st">"_genome_chr22_1_25409234"</span>,<span class="st">""</span>,<span class="fu">colData</span>(<span class="no">seExtended</span>)$<span class="no">name</span>)
<span class="fu"><a href="../reference/plot.bambu.html">plot.bambu</a></span>(<span class="no">seExtended</span>, <span class="kw">group.variable</span> <span class="kw">=</span> <span class="st">"groupVar"</span>, <span class="kw">type</span> <span class="kw">=</span> <span class="st">"heatmap"</span>)</pre></body></html></div>
<p><img src="bambu_files/figure-html/unnamed-chunk-8-1.png" width="768"></p>
<p>or with PCA plot</p>
<div class="sourceCode"><html><body><pre class="r"><span class="fu">colData</span>(<span class="no">seExtended</span>)$<span class="no">groupVar</span> <span class="kw">&lt;-</span>  <span class="fu"><a href="https://rdrr.io/r/base/unlist.html">unlist</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(<span class="no">seExtended</span>),<span class="kw">function</span>(<span class="no">x</span>) <span class="fu"><a href="https://rdrr.io/r/base/unlist.html">unlist</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/strsplit.html">strsplit</a></span>(<span class="no">x</span>,<span class="st">"_"</span>))[<span class="fl">2</span>]))
<span class="fu"><a href="../reference/plot.bambu.html">plot.bambu</a></span>(<span class="no">seExtended</span>, <span class="kw">group.variable</span> <span class="kw">=</span> <span class="st">"groupVar"</span>, <span class="kw">type</span> <span class="kw">=</span> <span class="st">"pca"</span>)</pre></body></html></div>
<p><img src="bambu_files/figure-html/unnamed-chunk-9-1.png" width="768"></p>
<div id="check-for-gene-examples" class="section level3">
<h3 class="hasAnchor">
<a href="#check-for-gene-examples" class="anchor"></a>Check for gene examples</h3>
<p>Single gene examples can also be checked using plot functions from <em>bambu</em></p>
<div class="sourceCode"><html><body><pre class="r"><span class="fu"><a href="../reference/plot.bambu.html">plot.bambu</a></span>(<span class="no">seExtended</span>, <span class="kw">type</span> <span class="kw">=</span> <span class="st">"annotation"</span>, <span class="kw">gene_id</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span>(<span class="fu">rowData</span>(<span class="no">seExtended</span>)$<span class="no">GENEID</span>)[<span class="fl">10</span>])</pre></body></html></div>
<p><img src="bambu_files/figure-html/unnamed-chunk-10-1.png" width="768"></p>
<pre><code>##&gt; [[1]]
##&gt; TableGrob (3 x 1) "arrange": 3 grobs
##&gt;   z     cells    name                grob
##&gt; 1 1 (2-2,1-1) arrange      gtable[layout]
##&gt; 2 2 (3-3,1-1) arrange      gtable[layout]
##&gt; 3 3 (1-1,1-1) arrange text[GRID.text.210]</code></pre>
</div>
<div id="transcript-to-gene-expression" class="section level3">
<h3 class="hasAnchor">
<a href="#transcript-to-gene-expression" class="anchor"></a>Transcript to Gene expression</h3>
<p>Gene expression can be obtained from transcript expression using this function:</p>
<div class="sourceCode"><html><body><pre class="r"><span class="no">seGene</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/transcriptToGeneExpression.html">transcriptToGeneExpression</a></span>(<span class="no">seExtended</span>)
<span class="no">seGene</span>
<span class="co">##&gt; class: RangedSummarizedExperiment </span>
<span class="co">##&gt; dim: 887 6 </span>
<span class="co">##&gt; metadata(0):</span>
<span class="co">##&gt; assays(2): counts CPM</span>
<span class="co">##&gt; rownames(887): ENSG00000015475 ENSG00000040608 ... gene.98 gene.99</span>
<span class="co">##&gt; rowData names(2): GENEID newGeneClass</span>
<span class="co">##&gt; colnames(6): SGNex_K562_directcDNA_replicate1_run2</span>
<span class="co">##&gt;   SGNex_K562_directcDNA_replicate4_run2 ...</span>
<span class="co">##&gt;   SGNex_MCF7_directcDNA_replicate3_run3</span>
<span class="co">##&gt;   SGNex_MCF7_directRNA_replicate4_run1</span>
<span class="co">##&gt; colData names(2): name groupVar</span></pre></body></html></div>
<p>Gene expression heatmap</p>
<div class="sourceCode"><html><body><pre class="r"><span class="fu">colData</span>(<span class="no">seGene</span>)$<span class="no">groupVar</span> <span class="kw">&lt;-</span>  <span class="fu"><a href="https://rdrr.io/r/base/unlist.html">unlist</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(<span class="no">seGene</span>),<span class="kw">function</span>(<span class="no">x</span>) <span class="fu"><a href="https://rdrr.io/r/base/unlist.html">unlist</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/strsplit.html">strsplit</a></span>(<span class="no">x</span>,<span class="st">"_"</span>))[<span class="fl">2</span>]))
<span class="fu"><a href="../reference/plot.bambu.html">plot.bambu</a></span>(<span class="no">seGene</span>, <span class="kw">group.variable</span> <span class="kw">=</span> <span class="st">"groupVar"</span>, <span class="kw">type</span> <span class="kw">=</span> <span class="st">"heatmap"</span>)</pre></body></html></div>
<p><img src="bambu_files/figure-html/unnamed-chunk-12-1.png" width="768"></p>
<p>Gene expression PCA plot</p>
<div class="sourceCode"><html><body><pre class="r"><span class="fu">colData</span>(<span class="no">seGene</span>)$<span class="no">groupVar</span> <span class="kw">&lt;-</span>  <span class="fu"><a href="https://rdrr.io/r/base/unlist.html">unlist</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(<span class="no">seGene</span>),<span class="kw">function</span>(<span class="no">x</span>) <span class="fu"><a href="https://rdrr.io/r/base/unlist.html">unlist</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/strsplit.html">strsplit</a></span>(<span class="no">x</span>,<span class="st">"_"</span>))[<span class="fl">2</span>]))
<span class="fu"><a href="../reference/plot.bambu.html">plot.bambu</a></span>(<span class="no">seGene</span>, <span class="kw">group.variable</span> <span class="kw">=</span> <span class="st">"groupVar"</span>, <span class="kw">type</span> <span class="kw">=</span> <span class="st">"pca"</span>)</pre></body></html></div>
<p><img src="bambu_files/figure-html/unnamed-chunk-13-1.png" width="768"></p>
</div>
</div>
<div id="differentially-expressed-genes" class="section level2">
<h2 class="hasAnchor">
<a href="#differentially-expressed-genes" class="anchor"></a>Differentially expressed genes</h2>
<p>We used <em>DESeq2</em> to find the differentially expressed genes:</p>
<div class="sourceCode"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">DESeq2</span>)
<span class="no">dds</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/DESeqDataSet.html">DESeqDataSetFromMatrix</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/apply.html">apply</a></span>(<span class="fu">assays</span>(<span class="no">seGene</span>)$<span class="no">counts</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">1</span>,<span class="fl">2</span>),<span class="no">round</span>),<span class="co">#tmp_wide[,-1],</span>
                                   <span class="kw">colData</span> <span class="kw">=</span> <span class="fu">colData</span>(<span class="no">seExtended</span>),
                                   <span class="kw">design</span> <span class="kw">=</span> ~ <span class="no">groupVar</span>)
<span class="fu"><a href="https://rdrr.io/r/base/system.time.html">system.time</a></span>(<span class="no">dds.deseq</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/DESeq.html">DESeq</a></span>(<span class="no">dds</span>))
<span class="co">##&gt;    user  system elapsed </span>
<span class="co">##&gt;   2.092   0.000   2.090</span>

<span class="no">deGeneRes</span> <span class="kw">&lt;-</span> <span class="kw pkg">DESeq2</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/results.html">results</a></span>(<span class="no">dds.deseq</span>, <span class="kw">independentFiltering</span><span class="kw">=</span><span class="fl">FALSE</span>)</pre></body></html></div>
<div class="sourceCode"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="no">deGeneRes</span>[<span class="fu"><a href="https://rdrr.io/r/base/order.html">order</a></span>(<span class="no">deGeneRes</span>$<span class="no">padj</span>),])
<span class="co">##&gt; log2 fold change (MLE): groupVar MCF7 vs K562 </span>
<span class="co">##&gt; Wald test p-value: groupVar MCF7 vs K562 </span>
<span class="co">##&gt; DataFrame with 6 rows and 6 columns</span>
<span class="co">##&gt;                  baseMean log2FoldChange     lfcSE      stat      pvalue</span>
<span class="co">##&gt;                 &lt;numeric&gt;      &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;   &lt;numeric&gt;</span>
<span class="co">##&gt; ENSG00000185686  513.5554       -7.22327  0.498140 -14.50048 1.20303e-47</span>
<span class="co">##&gt; ENSG00000197077   26.2357        9.09835  1.326114   6.86091 6.84254e-12</span>
<span class="co">##&gt; ENSG00000283633   88.7407       -9.03039  1.431724  -6.30736 2.83841e-10</span>
<span class="co">##&gt; ENSG00000099977  232.6720        1.85141  0.306629   6.03795 1.56080e-09</span>
<span class="co">##&gt; ENSG00000169635   44.4275       -3.44047  0.579080  -5.94127 2.82821e-09</span>
<span class="co">##&gt; ENSG00000100181   38.3698       -5.07047  0.892940  -5.67840 1.35959e-08</span>
<span class="co">##&gt;                        padj</span>
<span class="co">##&gt;                   &lt;numeric&gt;</span>
<span class="co">##&gt; ENSG00000185686 7.14599e-45</span>
<span class="co">##&gt; ENSG00000197077 2.03224e-09</span>
<span class="co">##&gt; ENSG00000283633 5.62005e-08</span>
<span class="co">##&gt; ENSG00000099977 2.31779e-07</span>
<span class="co">##&gt; ENSG00000169635 3.35991e-07</span>
<span class="co">##&gt; ENSG00000100181 1.34599e-06</span></pre></body></html></div>
<div class="sourceCode"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/summary.html">summary</a></span>(<span class="no">deGeneRes</span>)
<span class="co">##&gt; </span>
<span class="co">##&gt; out of 594 with nonzero total read count</span>
<span class="co">##&gt; adjusted p-value &lt; 0.1</span>
<span class="co">##&gt; LFC &gt; 0 (up)       : 16, 2.7%</span>
<span class="co">##&gt; LFC &lt; 0 (down)     : 20, 3.4%</span>
<span class="co">##&gt; outliers [1]       : 0, 0%</span>
<span class="co">##&gt; low counts [2]     : 0, 0%</span>
<span class="co">##&gt; (mean count &lt; 0)</span>
<span class="co">##&gt; [1] see 'cooksCutoff' argument of ?results</span>
<span class="co">##&gt; [2] see 'independentFiltering' argument of ?results</span></pre></body></html></div>
<div class="sourceCode"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/plotMA.html">plotMA</a></span>(<span class="no">deGeneRes</span>, <span class="kw">ylim</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(-<span class="fl">3</span>,<span class="fl">3</span>))</pre></body></html></div>
<p><img src="bambu_files/figure-html/unnamed-chunk-17-1.png" width="768"></p>
<p>Plotting shrinked lFC results</p>
<div class="sourceCode"><html><body><pre class="r"><span class="no">resLFC</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/lfcShrink.html">lfcShrink</a></span>(<span class="no">dds.deseq</span>, <span class="kw">coef</span><span class="kw">=</span><span class="st">"groupVar_MCF7_vs_K562"</span>, <span class="kw">type</span><span class="kw">=</span><span class="st">"apeglm"</span>)
<span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/plotMA.html">plotMA</a></span>(<span class="no">resLFC</span>, <span class="kw">ylim</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(-<span class="fl">3</span>,<span class="fl">3</span>))</pre></body></html></div>
<p><img src="bambu_files/figure-html/unnamed-chunk-18-1.png" width="768"></p>
</div>
<div id="differential-expression-for-isoform-detection" class="section level2">
<h2 class="hasAnchor">
<a href="#differential-expression-for-isoform-detection" class="anchor"></a>Differential expression for isoform detection</h2>
<p>We used <em>DEXSeq</em> to detect alternative used isoforms.</p>
<div class="sourceCode"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">DRIMSeq</span>)
<span class="no">count.data</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span>(<span class="fu">rowData</span>(<span class="no">seExtended</span>))
<span class="no">count.data</span>$<span class="no">gene_id</span> <span class="kw">&lt;-</span> <span class="no">count.data</span>$<span class="no">GENEID</span>
<span class="no">count.data</span>$<span class="no">feature_id</span> <span class="kw">&lt;-</span> <span class="no">count.data</span>$<span class="no">TXNAME</span>
<span class="no">count.data</span>$<span class="no">GENEID</span> <span class="kw">&lt;-</span> <span class="no">count.data</span>$<span class="no">TXNAME</span> <span class="kw">&lt;-</span> <span class="kw">NULL</span>

<span class="no">count.data</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html">cbind</a></span>(<span class="no">count.data</span>, <span class="fu">assays</span>(<span class="no">seExtended</span>)$<span class="no">counts</span>)

<span class="no">sample.info</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span>(<span class="fu">colData</span>(<span class="no">seExtended</span>))
<span class="no">sample.info</span>$<span class="no">sample_id</span> <span class="kw">&lt;-</span> <span class="no">sample.info</span>$<span class="no">name</span>
<span class="no">sample.info</span>$<span class="no">name</span> <span class="kw">&lt;-</span> <span class="kw">NULL</span>
<span class="no">d</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DRIMSeq/man/dmDSdata.html">dmDSdata</a></span>(<span class="kw">counts</span><span class="kw">=</span><span class="no">count.data</span>, <span class="kw">samples</span><span class="kw">=</span><span class="no">sample.info</span>)

<span class="no">n_samp_gene</span> <span class="kw">&lt;-</span> <span class="fl">1</span>
<span class="no">n_samp_feature</span> <span class="kw">&lt;-</span> <span class="fl">1</span>
<span class="no">min_count_gene</span> <span class="kw">&lt;-</span> <span class="fl">1</span>
<span class="no">min_count_feature</span> <span class="kw">&lt;-</span> <span class="fl">1</span>
<span class="no">dFilter</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DRIMSeq/man/dmFilter.html">dmFilter</a></span>(<span class="no">d</span>,
              <span class="kw">min_samps_feature_expr</span> <span class="kw">=</span> <span class="no">n_samp_feature</span>,
              <span class="kw">min_samps_feature_prop</span> <span class="kw">=</span> <span class="no">n_samp_feature</span>,
              <span class="kw">min_samps_gene_expr</span> <span class="kw">=</span> <span class="no">n_samp_gene</span>,
              <span class="kw">min_feature_expr</span> <span class="kw">=</span> <span class="no">min_count_feature</span>,
              <span class="kw">min_gene_expr</span> <span class="kw">=</span> <span class="no">min_count_gene</span>,
              <span class="kw">min_feature_prop</span><span class="kw">=</span><span class="fl">0.1</span>)
<span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span>(<span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/counts.html">counts</a></span>(<span class="no">dFilter</span>)$<span class="no">gene_id</span>)) <span class="co">## number of isoforms </span>
<span class="co">##&gt; </span>
<span class="co">##&gt;  2  3  4  5  6  7  8 10 </span>
<span class="co">##&gt; 23 15 21 12  8  6  4  1</span></pre></body></html></div>
<div class="sourceCode"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">DEXSeq</span>)
<span class="no">formulaFullModel</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/formula.html">as.formula</a></span>(<span class="st">"~sample + exon + groupVar:exon"</span>)


<span class="no">dxd</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DEXSeq/man/DEXSeqDataSet.html">DEXSeqDataSet</a></span>(<span class="kw">countData</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/Round.html">round</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span>(<span class="fu"><a href="https://rdrr.io/pkg/DEXSeq/man/counts.html">counts</a></span>(<span class="no">dFilter</span>)[,-<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">1</span>:<span class="fl">2</span>)])),
                     <span class="kw">sampleData</span><span class="kw">=</span><span class="kw pkg">DRIMSeq</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/DRIMSeq/man/dmDSdata-class.html">samples</a></span>(<span class="no">dFilter</span>),
                     <span class="kw">design</span><span class="kw">=</span><span class="no">formulaFullModel</span>,
                     <span class="kw">featureID</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/pkg/DEXSeq/man/counts.html">counts</a></span>(<span class="no">dFilter</span>)$<span class="no">feature_id</span>,
                     <span class="kw">groupID</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/pkg/DEXSeq/man/counts.html">counts</a></span>(<span class="no">dFilter</span>)$<span class="no">gene_id</span>)


<span class="fu"><a href="https://rdrr.io/r/base/system.time.html">system.time</a></span>({
  <span class="no">dxd</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DEXSeq/man/estimateSizeFactors.html">estimateSizeFactors</a></span>(<span class="no">dxd</span>)
  <span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span>(<span class="st">'Size factor estimated'</span>)
  <span class="no">dxd</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DEXSeq/man/estimateDispersions.html">estimateDispersions</a></span>(<span class="no">dxd</span>, <span class="kw">formula</span> <span class="kw">=</span> <span class="no">formulaFullModel</span>)
  <span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span>(<span class="st">'Dispersion estimated'</span>)
  <span class="co">#dxd &lt;- estimateExonFoldChanges( dxd )</span>
  <span class="no">dxd</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DEXSeq/man/testForDEU.html">testForDEU</a></span>(<span class="no">dxd</span>, <span class="kw">fullModel</span> <span class="kw">=</span> <span class="no">formulaFullModel</span>)
  <span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span>(<span class="st">'DEU tested'</span>)
  <span class="no">dxd</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DEXSeq/man/estimateExonFoldChanges.html">estimateExonFoldChanges</a></span>(<span class="no">dxd</span>, <span class="kw">fitExpToVar</span><span class="kw">=</span><span class="st">"groupVar"</span>)
  <span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span>(<span class="st">'Exon fold changes estimated'</span>)
})
<span class="co">##&gt; [1] "Size factor estimated"</span>
<span class="co">##&gt; [1] "Dispersion estimated"</span>
<span class="co">##&gt; [1] "DEU tested"</span>
<span class="co">##&gt; [1] "Exon fold changes estimated"</span>
<span class="co">##&gt;    user  system elapsed </span>
<span class="co">##&gt;   8.676   0.044   8.719</span></pre></body></html></div>
<div class="sourceCode"><html><body><pre class="r"><span class="no">dxr</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DEXSeq/man/DEXSeqResults.html">DEXSeqResults</a></span>(<span class="no">dxd</span>, <span class="kw">independentFiltering</span><span class="kw">=</span><span class="fl">FALSE</span>)
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="no">dxr</span>)
<span class="co">##&gt; </span>
<span class="co">##&gt; LRT p-value: full vs reduced</span>
<span class="co">##&gt; </span>
<span class="co">##&gt; DataFrame with 6 rows and 12 columns</span>
<span class="co">##&gt;                   groupID   featureID exonBaseMean dispersion      stat</span>
<span class="co">##&gt;               &lt;character&gt; &lt;character&gt;    &lt;numeric&gt;  &lt;numeric&gt; &lt;numeric&gt;</span>
<span class="co">##&gt; gene.2:tx.2        gene.2        tx.2     2.714822   0.427670  3.150551</span>
<span class="co">##&gt; gene.2:tx.3        gene.2        tx.3     0.654029   0.791030  0.140914</span>
<span class="co">##&gt; gene.2:tx.4        gene.2        tx.4     2.567833   0.315359  1.003001</span>
<span class="co">##&gt; gene.2:tx.66       gene.2       tx.66     2.107202   0.332461  0.422016</span>
<span class="co">##&gt; gene.2:tx.67       gene.2       tx.67     1.243100   1.748445  1.683733</span>
<span class="co">##&gt; gene.2:tx.122      gene.2      tx.122     1.865452   0.467004  1.043483</span>
<span class="co">##&gt;                  pvalue      padj      K562      MCF7 log2fold_MCF7_K562</span>
<span class="co">##&gt;               &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;          &lt;numeric&gt;</span>
<span class="co">##&gt; gene.2:tx.2   0.0759013   0.99485   2.14944 0.0448427         -11.620740</span>
<span class="co">##&gt; gene.2:tx.3   0.7073738   1.00000   1.15465 0.0448419          -9.506951</span>
<span class="co">##&gt; gene.2:tx.4   0.3165853   1.00000   2.03209 1.1710398          -1.860342</span>
<span class="co">##&gt; gene.2:tx.66  0.5159328   1.00000   1.58048 1.8911532           0.622989</span>
<span class="co">##&gt; gene.2:tx.67  0.1944292   1.00000   1.05916 1.7073365           1.554900</span>
<span class="co">##&gt; gene.2:tx.122 0.3070128   1.00000   1.44728 1.8878239           0.910422</span>
<span class="co">##&gt;                 genomicData   countData</span>
<span class="co">##&gt;               &lt;GRangesList&gt;    &lt;matrix&gt;</span>
<span class="co">##&gt; gene.2:tx.2                 18:13:3:...</span>
<span class="co">##&gt; gene.2:tx.3                   2:3:2:...</span>
<span class="co">##&gt; gene.2:tx.4                 10:15:3:...</span>
<span class="co">##&gt; gene.2:tx.66                  7:5:3:...</span>
<span class="co">##&gt; gene.2:tx.67                  0:6:0:...</span>
<span class="co">##&gt; gene.2:tx.122                 2:9:1:...</span></pre></body></html></div>
<div class="sourceCode"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">stageR</span>)
<span class="no">strp</span> <span class="kw">&lt;-</span> <span class="kw">function</span>(<span class="no">x</span>) <span class="fu"><a href="https://rdrr.io/r/base/substr.html">substr</a></span>(<span class="no">x</span>,<span class="fl">1</span>,<span class="fl">15</span>)
<span class="no">qval</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DEXSeq/man/perGeneQValue.html">perGeneQValue</a></span>(<span class="no">dxr</span>)
<span class="no">dxr.g</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span>(<span class="kw">gene</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span>(<span class="no">qval</span>),<span class="no">qval</span>)

<span class="no">columns</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"featureID"</span>,<span class="st">"groupID"</span>,<span class="st">"pvalue"</span>)
<span class="no">dxr_pval</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span>(<span class="no">dxr</span>[,<span class="no">columns</span>])
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="no">dxr_pval</span>)
<span class="co">##&gt;               featureID groupID     pvalue</span>
<span class="co">##&gt; gene.2:tx.2        tx.2  gene.2 0.07590133</span>
<span class="co">##&gt; gene.2:tx.3        tx.3  gene.2 0.70737384</span>
<span class="co">##&gt; gene.2:tx.4        tx.4  gene.2 0.31658535</span>
<span class="co">##&gt; gene.2:tx.66      tx.66  gene.2 0.51593283</span>
<span class="co">##&gt; gene.2:tx.67      tx.67  gene.2 0.19442922</span>
<span class="co">##&gt; gene.2:tx.122    tx.122  gene.2 0.30701285</span>

<span class="no">pConfirmation</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span>(<span class="no">dxr_pval</span>$<span class="no">pvalue</span>,<span class="kw">ncol</span><span class="kw">=</span><span class="fl">1</span>)
<span class="fu"><a href="https://rdrr.io/r/base/dimnames.html">dimnames</a></span>(<span class="no">pConfirmation</span>) <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="fu">strp</span>(<span class="no">dxr_pval</span>$<span class="no">featureID</span>),<span class="st">"transcript"</span>)
<span class="no">pScreen</span> <span class="kw">&lt;-</span> <span class="no">qval</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span>(<span class="no">pScreen</span>) <span class="kw">&lt;-</span> <span class="fu">strp</span>(<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span>(<span class="no">pScreen</span>))
<span class="no">tx2gene</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span>(<span class="no">dxr_pval</span>[,<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"featureID"</span>, <span class="st">"groupID"</span>)])
<span class="kw">for</span> (<span class="no">i</span> <span class="kw">in</span> <span class="fl">1</span>:<span class="fl">2</span>) <span class="no">tx2gene</span>[,<span class="no">i</span>] <span class="kw">&lt;-</span> <span class="fu">strp</span>(<span class="no">tx2gene</span>[,<span class="no">i</span>])


<span class="no">stageRObj</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/stageR/man/stageRTx.html">stageRTx</a></span>(<span class="kw">pScreen</span><span class="kw">=</span><span class="no">pScreen</span>, <span class="kw">pConfirmation</span><span class="kw">=</span><span class="no">pConfirmation</span>,
                      <span class="kw">pScreenAdjusted</span><span class="kw">=</span><span class="fl">TRUE</span>, <span class="kw">tx2gene</span><span class="kw">=</span><span class="no">tx2gene</span>)
<span class="no">stageRObj</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/stageR/man/stageWiseAdjustment.html">stageWiseAdjustment</a></span>(<span class="no">stageRObj</span>, <span class="kw">method</span><span class="kw">=</span><span class="st">"dtu"</span>, <span class="kw">alpha</span><span class="kw">=</span><span class="fl">0.5</span>)
<span class="fu"><a href="https://rdrr.io/r/base/warning.html">suppressWarnings</a></span>({
  <span class="no">dex.padj</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/stageR/man/getAdjustedPValues.html">getAdjustedPValues</a></span>(<span class="no">stageRObj</span>, <span class="kw">order</span><span class="kw">=</span><span class="fl">FALSE</span>,
                                 <span class="kw">onlySignificantGenes</span><span class="kw">=</span><span class="fl">TRUE</span>)
})</pre></body></html></div>
<div class="sourceCode"><html><body><pre class="r"><span class="no">dxrDT</span> <span class="kw">&lt;-</span> <span class="fu">data.table</span>(<span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span>(<span class="no">dxr</span>))
<span class="fu">setnames</span>(<span class="no">dxrDT</span>, <span class="kw">old</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">'groupID'</span>,<span class="st">'featureID'</span>), <span class="kw">new</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">'geneID'</span>,<span class="st">'txID'</span>))
<span class="no">dex.padj</span> <span class="kw">&lt;-</span> <span class="fu">data.table</span>(<span class="no">dex.padj</span>)

<span class="no">dxrDT</span> <span class="kw">&lt;-</span> <span class="no">dex.padj</span>[<span class="no">dxrDT</span>, <span class="kw">on</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">'geneID'</span>,<span class="st">'txID'</span>)]
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="no">dxrDT</span>)
<span class="co">##&gt;    geneID   txID gene transcript exonBaseMean dispersion      stat     pvalue</span>
<span class="co">##&gt; 1: gene.2   tx.2   NA         NA    2.7148218  0.4276702 3.1505509 0.07590133</span>
<span class="co">##&gt; 2: gene.2   tx.3   NA         NA    0.6540286  0.7910296 0.1409142 0.70737384</span>
<span class="co">##&gt; 3: gene.2   tx.4   NA         NA    2.5678325  0.3153589 1.0030014 0.31658535</span>
<span class="co">##&gt; 4: gene.2  tx.66   NA         NA    2.1072015  0.3324607 0.4220160 0.51593283</span>
<span class="co">##&gt; 5: gene.2  tx.67   NA         NA    1.2431000  1.7484453 1.6837326 0.19442922</span>
<span class="co">##&gt; 6: gene.2 tx.122   NA         NA    1.8654524  0.4670041 1.0434828 0.30701285</span>
<span class="co">##&gt;         padj     K562       MCF7 log2fold_MCF7_K562 genomicData</span>
<span class="co">##&gt; 1: 0.9948496 2.149443 0.04484265        -11.6207395   &lt;GRanges&gt;</span>
<span class="co">##&gt; 2: 1.0000000 1.154654 0.04484192         -9.5069512   &lt;GRanges&gt;</span>
<span class="co">##&gt; 3: 1.0000000 2.032089 1.17103980         -1.8603417   &lt;GRanges&gt;</span>
<span class="co">##&gt; 4: 1.0000000 1.580476 1.89115320          0.6229887   &lt;GRanges&gt;</span>
<span class="co">##&gt; 5: 1.0000000 1.059156 1.70733655          1.5549004   &lt;GRanges&gt;</span>
<span class="co">##&gt; 6: 1.0000000 1.447282 1.88782389          0.9104218   &lt;GRanges&gt;</span>
<span class="co">##&gt;    countData.SGNex_K562_directcDNA_replicate1_run2</span>
<span class="co">##&gt; 1:                                              18</span>
<span class="co">##&gt; 2:                                               2</span>
<span class="co">##&gt; 3:                                              10</span>
<span class="co">##&gt; 4:                                               7</span>
<span class="co">##&gt; 5:                                               0</span>
<span class="co">##&gt; 6:                                               2</span>
<span class="co">##&gt;    countData.SGNex_K562_directcDNA_replicate4_run2</span>
<span class="co">##&gt; 1:                                              13</span>
<span class="co">##&gt; 2:                                               3</span>
<span class="co">##&gt; 3:                                              15</span>
<span class="co">##&gt; 4:                                               5</span>
<span class="co">##&gt; 5:                                               6</span>
<span class="co">##&gt; 6:                                               9</span>
<span class="co">##&gt;    countData.SGNex_K562_directRNA_replicate6_run1</span>
<span class="co">##&gt; 1:                                              3</span>
<span class="co">##&gt; 2:                                              2</span>
<span class="co">##&gt; 3:                                              3</span>
<span class="co">##&gt; 4:                                              3</span>
<span class="co">##&gt; 5:                                              0</span>
<span class="co">##&gt; 6:                                              1</span>
<span class="co">##&gt;    countData.SGNex_MCF7_directcDNA_replicate1_run2</span>
<span class="co">##&gt; 1:                                               0</span>
<span class="co">##&gt; 2:                                               0</span>
<span class="co">##&gt; 3:                                               0</span>
<span class="co">##&gt; 4:                                               1</span>
<span class="co">##&gt; 5:                                               1</span>
<span class="co">##&gt; 6:                                               1</span>
<span class="co">##&gt;    countData.SGNex_MCF7_directcDNA_replicate3_run3</span>
<span class="co">##&gt; 1:                                               0</span>
<span class="co">##&gt; 2:                                               0</span>
<span class="co">##&gt; 3:                                               1</span>
<span class="co">##&gt; 4:                                               2</span>
<span class="co">##&gt; 5:                                               0</span>
<span class="co">##&gt; 6:                                               2</span>
<span class="co">##&gt;    countData.SGNex_MCF7_directRNA_replicate4_run1</span>
<span class="co">##&gt; 1:                                              0</span>
<span class="co">##&gt; 2:                                              0</span>
<span class="co">##&gt; 3:                                              0</span>
<span class="co">##&gt; 4:                                              0</span>
<span class="co">##&gt; 5:                                              1</span>
<span class="co">##&gt; 6:                                              0</span></pre></body></html></div>
<div class="sourceCode"><html><body><pre class="r"><span class="no">dxrDT</span>[,<span class="no">sigLFC2</span><span class="kw">:=</span>(<span class="no">padj</span> <span class="kw">&lt;</span> <span class="fl">0.5</span><span class="kw">&amp;</span>(<span class="fu"><a href="https://rdrr.io/r/base/MathFun.html">abs</a></span>(<span class="no">log2fold_MCF7_K562</span>)<span class="kw">&gt;=</span><span class="fl">2</span>))]
<span class="fu">ggplot</span>(<span class="no">dxrDT</span>, <span class="fu">aes</span>(<span class="kw">y</span> <span class="kw">=</span> <span class="no">log2fold_MCF7_K562</span>, <span class="kw">x</span> <span class="kw">=</span> <span class="no">exonBaseMean</span>, <span class="kw">color</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">as.factor</a></span>(<span class="no">padj</span><span class="kw">&lt;</span><span class="fl">0.5</span>)))+
  <span class="fu">geom_point</span>(<span class="kw">size</span> <span class="kw">=</span> <span class="fl">0.5</span>)+
     <span class="fu">scale_x_log10</span>()+
  <span class="fu">scale_color_manual</span>(<span class="kw">values</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">'grey'</span>,<span class="st">'indianred'</span>), <span class="kw">name</span> <span class="kw">=</span> <span class="st">"Significant"</span>)+
     <span class="fu">xlab</span>(<span class="st">"Mean of normalized counts"</span>)+
    <span class="fu">ylab</span>(<span class="st">"Log2 Fold change"</span>)+
  <span class="fu">theme_minimal</span>()</pre></body></html></div>
<p><img src="bambu_files/figure-html/unnamed-chunk-24-1.png" width="768"></p>
<div class="sourceCode"><html><body><pre class="r"><span class="no">dxrDT</span>[<span class="no">padj</span><span class="kw">&lt;</span><span class="fl">0.5</span>,<span class="fu">.</span>(<span class="no">geneID</span>, <span class="no">txID</span>, <span class="no">log2fold_MCF7_K562</span>,<span class="no">K562</span>,<span class="no">MCF7</span>)]
<span class="co">##&gt;             geneID            txID log2fold_MCF7_K562      K562     MCF7</span>
<span class="co">##&gt; 1: ENSG00000183597           tx.35          -2.925745 4.2504114 2.284457</span>
<span class="co">##&gt; 2: ENSG00000184436           tx.44          -4.132398 3.3428343 1.103626</span>
<span class="co">##&gt; 3: ENSG00000272779          tx.105           4.276779 0.6241729 2.320295</span>
<span class="co">##&gt; 4: ENSG00000100030 ENST00000215832           1.112208 5.7031758 6.535581</span>
<span class="co">##&gt; 5: ENSG00000100030 ENST00000398822          -1.438071 4.0021766 3.004256</span>
<span class="co">##&gt; 6: ENSG00000169635 ENST00000407464           2.362562 1.9298346 3.404064</span>
<span class="co">##&gt; 7: ENSG00000169635 ENST00000443632          -1.904184 4.0971165 2.785404</span>
<span class="co">##&gt; 8: ENSG00000215012 ENST00000407472          -1.753270 3.6246114 2.462383</span></pre></body></html></div>
<div class="sourceCode"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span>()
<span class="co">##&gt; R version 4.0.0 (2020-04-24)</span>
<span class="co">##&gt; Platform: x86_64-pc-linux-gnu (64-bit)</span>
<span class="co">##&gt; Running under: Ubuntu 16.04.6 LTS</span>
<span class="co">##&gt; </span>
<span class="co">##&gt; Matrix products: default</span>
<span class="co">##&gt; BLAS:   /usr/lib/libblas/libblas.so.3.6.0</span>
<span class="co">##&gt; LAPACK: /usr/lib/lapack/liblapack.so.3.6.0</span>
<span class="co">##&gt; </span>
<span class="co">##&gt; locale:</span>
<span class="co">##&gt;  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span>
<span class="co">##&gt;  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span>
<span class="co">##&gt;  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span>
<span class="co">##&gt;  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span>
<span class="co">##&gt;  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span>
<span class="co">##&gt; [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span>
<span class="co">##&gt; </span>
<span class="co">##&gt; attached base packages:</span>
<span class="co">##&gt; [1] parallel  stats4    stats     graphics  grDevices utils     datasets </span>
<span class="co">##&gt; [8] methods   base     </span>
<span class="co">##&gt; </span>
<span class="co">##&gt; other attached packages:</span>
<span class="co">##&gt;  [1] stageR_1.10.0               DEXSeq_1.34.0              </span>
<span class="co">##&gt;  [3] RColorBrewer_1.1-2          BiocParallel_1.22.0        </span>
<span class="co">##&gt;  [5] DRIMSeq_1.16.0              DESeq2_1.28.1              </span>
<span class="co">##&gt;  [7] NanoporeRNASeq_0.9.0        bambu_0.1.0                </span>
<span class="co">##&gt;  [9] testthat_2.3.2              ggplot2_3.3.0              </span>
<span class="co">##&gt; [11] GenomicFeatures_1.40.0      AnnotationDbi_1.50.0       </span>
<span class="co">##&gt; [13] BiocManager_1.30.10         SummarizedExperiment_1.18.1</span>
<span class="co">##&gt; [15] DelayedArray_0.14.0         matrixStats_0.56.0         </span>
<span class="co">##&gt; [17] Biobase_2.48.0              GenomicRanges_1.40.0       </span>
<span class="co">##&gt; [19] GenomeInfoDb_1.24.0         IRanges_2.22.1             </span>
<span class="co">##&gt; [21] S4Vectors_0.26.1            BiocGenerics_0.34.0        </span>
<span class="co">##&gt; [23] dplyr_0.8.5                 data.table_1.12.8          </span>
<span class="co">##&gt; </span>
<span class="co">##&gt; loaded via a namespace (and not attached):</span>
<span class="co">##&gt;   [1] backports_1.1.7           circlize_0.4.9           </span>
<span class="co">##&gt;   [3] Hmisc_4.4-0               BiocFileCache_1.12.0     </span>
<span class="co">##&gt;   [5] plyr_1.8.6                lazyeval_0.2.2           </span>
<span class="co">##&gt;   [7] splines_4.0.0             usethis_1.6.1            </span>
<span class="co">##&gt;   [9] digest_0.6.25             ensembldb_2.12.1         </span>
<span class="co">##&gt;  [11] foreach_1.5.0             htmltools_0.4.0          </span>
<span class="co">##&gt;  [13] fansi_0.4.1               magrittr_1.5             </span>
<span class="co">##&gt;  [15] checkmate_2.0.0           memoise_1.1.0            </span>
<span class="co">##&gt;  [17] BSgenome_1.56.0           cluster_2.1.0            </span>
<span class="co">##&gt;  [19] limma_3.44.1              remotes_2.1.1            </span>
<span class="co">##&gt;  [21] annotate_1.66.0           ComplexHeatmap_2.4.2     </span>
<span class="co">##&gt;  [23] Biostrings_2.56.0         ggbio_1.36.0             </span>
<span class="co">##&gt;  [25] bdsmatrix_1.3-4           askpass_1.1              </span>
<span class="co">##&gt;  [27] pkgdown_1.5.1             prettyunits_1.1.1        </span>
<span class="co">##&gt;  [29] jpeg_0.1-8.1              colorspace_1.4-1         </span>
<span class="co">##&gt;  [31] apeglm_1.10.0             blob_1.2.1               </span>
<span class="co">##&gt;  [33] rappdirs_0.3.1            xfun_0.14                </span>
<span class="co">##&gt;  [35] callr_3.4.3               crayon_1.3.4             </span>
<span class="co">##&gt;  [37] RCurl_1.98-1.2            RcppArmadillo_0.9.880.1.0</span>
<span class="co">##&gt;  [39] graph_1.66.0              genefilter_1.70.0        </span>
<span class="co">##&gt;  [41] VariantAnnotation_1.34.0  survival_3.1-12          </span>
<span class="co">##&gt;  [43] iterators_1.0.12          glue_1.4.1               </span>
<span class="co">##&gt;  [45] gtable_0.3.0              zlibbioc_1.34.0          </span>
<span class="co">##&gt;  [47] XVector_0.28.0            GetoptLong_0.1.8         </span>
<span class="co">##&gt;  [49] pkgbuild_1.0.8            shape_1.4.4              </span>
<span class="co">##&gt;  [51] scales_1.1.1              mvtnorm_1.1-0            </span>
<span class="co">##&gt;  [53] edgeR_3.30.0              DBI_1.1.0                </span>
<span class="co">##&gt;  [55] GGally_1.5.0              Rcpp_1.0.4.6             </span>
<span class="co">##&gt;  [57] emdbook_1.3.12            xtable_1.8-4             </span>
<span class="co">##&gt;  [59] progress_1.2.2            htmlTable_1.13.3         </span>
<span class="co">##&gt;  [61] clue_0.3-57               foreign_0.8-79           </span>
<span class="co">##&gt;  [63] bit_1.1-15.2              OrganismDbi_1.30.0       </span>
<span class="co">##&gt;  [65] Formula_1.2-3             glmnet_4.0               </span>
<span class="co">##&gt;  [67] htmlwidgets_1.5.1         httr_1.4.1               </span>
<span class="co">##&gt;  [69] acepack_1.4.1             ellipsis_0.3.1           </span>
<span class="co">##&gt;  [71] pkgconfig_2.0.3           reshape_0.8.8            </span>
<span class="co">##&gt;  [73] XML_3.99-0.3              farver_2.0.3             </span>
<span class="co">##&gt;  [75] nnet_7.3-14               dbplyr_1.4.3             </span>
<span class="co">##&gt;  [77] locfit_1.5-9.4            tidyselect_1.1.0         </span>
<span class="co">##&gt;  [79] labeling_0.3              rlang_0.4.6              </span>
<span class="co">##&gt;  [81] reshape2_1.4.4            munsell_0.5.0            </span>
<span class="co">##&gt;  [83] tools_4.0.0               cli_2.0.2                </span>
<span class="co">##&gt;  [85] RSQLite_2.2.0             devtools_2.3.0           </span>
<span class="co">##&gt;  [87] evaluate_0.14             stringr_1.4.0            </span>
<span class="co">##&gt;  [89] yaml_2.2.1                processx_3.4.2           </span>
<span class="co">##&gt;  [91] knitr_1.28                bit64_0.9-7              </span>
<span class="co">##&gt;  [93] fs_1.4.1                  purrr_0.3.4              </span>
<span class="co">##&gt;  [95] AnnotationFilter_1.12.0   RBGL_1.64.0              </span>
<span class="co">##&gt;  [97] biomaRt_2.44.0            compiler_4.0.0           </span>
<span class="co">##&gt;  [99] rstudioapi_0.11           curl_4.3                 </span>
<span class="co">##&gt; [101] png_0.1-7                 statmod_1.4.34           </span>
<span class="co">##&gt; [103] geneplotter_1.66.0        tibble_3.0.1             </span>
<span class="co">##&gt; [105] stringi_1.4.6             ps_1.3.3                 </span>
<span class="co">##&gt; [107] desc_1.2.0                lattice_0.20-41          </span>
<span class="co">##&gt; [109] ProtGenerics_1.20.0       Matrix_1.2-18            </span>
<span class="co">##&gt; [111] vctrs_0.3.0               pillar_1.4.4             </span>
<span class="co">##&gt; [113] lifecycle_0.2.0           GlobalOptions_0.1.1      </span>
<span class="co">##&gt; [115] bitops_1.0-6              rtracklayer_1.48.0       </span>
<span class="co">##&gt; [117] hwriter_1.3.2             R6_2.4.1                 </span>
<span class="co">##&gt; [119] latticeExtra_0.6-29       RcppProgress_0.4.2       </span>
<span class="co">##&gt; [121] gridExtra_2.3             sessioninfo_1.1.1        </span>
<span class="co">##&gt; [123] codetools_0.2-16          dichromat_2.0-0          </span>
<span class="co">##&gt; [125] MASS_7.3-51.6             assertthat_0.2.1         </span>
<span class="co">##&gt; [127] pkgload_1.0.2             openssl_1.4.1            </span>
<span class="co">##&gt; [129] rprojroot_1.3-2           rjson_0.2.20             </span>
<span class="co">##&gt; [131] withr_2.2.0               GenomicAlignments_1.24.0 </span>
<span class="co">##&gt; [133] Rsamtools_2.4.0           GenomeInfoDbData_1.2.3   </span>
<span class="co">##&gt; [135] hms_0.5.3                 grid_4.0.0               </span>
<span class="co">##&gt; [137] rpart_4.1-15              coda_0.19-3              </span>
<span class="co">##&gt; [139] rmarkdown_2.1             biovizBase_1.36.0        </span>
<span class="co">##&gt; [141] bbmle_1.0.23.1            numDeriv_2016.8-1.1      </span>
<span class="co">##&gt; [143] base64enc_0.1-3</span></pre></body></html></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Ying Chen Developer, Jonathan Goeke Developer.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
