<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>long read isoform reconstruction and quantification — bambu • bambu</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="long read isoform reconstruction and quantification — bambu" />
<meta property="og:description" content="This function takes bam file of genomic alignments and performs isoform recontruction and gene and transcript expression quantification.
It also allows saving of read class files of alignments, extending provided annotations, and quantification based on extended annotations.
When multiple samples are provided, extended annotations will be combined across samples to allow comparison." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">bambu</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/bambu.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/GoekeLab/bambu/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>long read isoform reconstruction and quantification</h1>
    <small class="dont-index">Source: <a href='https://github.com/GoekeLab/bambu/blob/master/R/bambu.R'><code>R/bambu.R</code></a></small>
    <div class="hidden name"><code>bambu.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This function takes bam file of genomic alignments and performs isoform recontruction and gene and transcript expression quantification.
It also allows saving of read class files of alignments, extending provided annotations, and quantification based on extended annotations.
When multiple samples are provided, extended annotations will be combined across samples to allow comparison.</p>
    </div>

    <pre class="usage"><span class='fu'>bambu</span>(
  <span class='kw'>reads</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>readClass.file</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>readClass.outputDir</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>annotations</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>genomeSequence</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>stranded</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
  <span class='kw'>ncore</span> <span class='kw'>=</span> <span class='fl'>1</span>,
  <span class='kw'>yieldSize</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>isoreParameters</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>emParameters</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>extendAnnotations</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
  <span class='kw'>verbose</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>
)</pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>reads</th>
      <td><p>A string or a vector of strings specifying the paths of bam files for genomic alignments, or a <code>BamFile</code> object or a <code>BamFileList</code>  object (see <code>Rsamtools</code>).</p></td>
    </tr>
    <tr>
      <th>readClass.file</th>
      <td><p>A string or a vector of strings specifying the read class files that are saved during previous run of <code>bambu</code>.</p></td>
    </tr>
    <tr>
      <th>readClass.outputDir</th>
      <td><p>A string variable specifying the path to where read class files will be saved.</p></td>
    </tr>
    <tr>
      <th>annotations</th>
      <td><p>A <code>TxDb</code> object or A GRangesList object obtained by <code><a href='prepareAnnotations.html'>prepareAnnotations</a></code> or <code><a href='prepareAnnotationsFromGTF.html'>prepareAnnotationsFromGTF</a></code>.</p></td>
    </tr>
    <tr>
      <th>genomeSequence</th>
      <td><p>A fasta file or a BSGenome object.</p></td>
    </tr>
    <tr>
      <th>ncore</th>
      <td><p>specifying number of cores used when parallel processing is used, defaults to 1.</p></td>
    </tr>
    <tr>
      <th>yieldSize</th>
      <td><p>see <code>Rsamtools</code>.</p></td>
    </tr>
    <tr>
      <th>isoreParameters</th>
      <td><p>A list of controlling parameters for isoform reconstruction process:</p><ul>
<li><p>prefix specifying prefix for new gene Ids (genePrefix.number), defaults to empty</p></li>
<li><p>remove.subsetTx indicating whether filter to remove read classes which are a subset of known transcripts(), defaults to TRUE</p></li>
<li><p>min.readCount specifying minimun read count to consider a read class valid in a sample, defaults to 2</p></li>
<li><p>min.readFractionByGene specifying minimum relative read count per gene, highly expressed genes will have many high read count low relative abundance transcripts that can be filtered, defaults to 0.05</p></li>
<li><p>min.sampleNumber specifying minimum sample number with minimum read count, defaults to 1</p></li>
<li><p>min.exonDistance specifying minum distance to known transcript to be considered valid as new, defaults to 35</p></li>
<li><p>min.exonOverlap specifying minimum number of bases shared with annotation to be assigned to the same gene id, defaults 10 base pairs</p></li>
</ul></td>
    </tr>
    <tr>
      <th>emParameters</th>
      <td><p>A list of controlling parameters for quantification algorithm estimation process:</p><ul>
<li><p>maxiter specifying maximum number of run interations, defaults to 10000.</p></li>
<li><p>bias specifying whether to correct for bias, defaults to FALSE.</p></li>
<li><p>conv specifying the covergence trheshold control, defaults to 0.0001.</p></li>
</ul></td>
    </tr>
    <tr>
      <th>extendAnnotations</th>
      <td><p>A logical variable indicating whether annotations are to be extended for quantification.</p></td>
    </tr>
    <tr>
      <th>verbose</th>
      <td><p>A logical variable indicating whether processing messages will be printed.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A list of two SummarizedExperiment object for transcript expression and gene expression.</p>
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>Main function</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Ying Chen Developer, Jonathan Goeke Developer.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


